### ISSUE TITLE
ENH: Manage Cloud URLs

### ISSUE THREAD

**Issue by remi-braun** [drivendataorg/cloudpathlib#152]
Hello, it's me again 😃 

Could it be possible to manage cloud URLs the same way than CloudPaths (ie. `https://my-bucket.s3.us-west-2.amazonaws.com/puppy.png`) ?

Or maybe just encapsulate them in a URLPath class to recognize them, because for now they are understood as PosixPaths, which can lead to some errors.

Thanks a lot !

**Comment by pjbull** [comment]
Do you need to parse the AWS cloud URLs into their constituent parts? IIRC, there are a fair number of different variations on these URLs that would need to be supported. Not sure the story on other cloud providers.

If this is just creating these URLs, #21 tracks producing them so they can be used, e.g. in a web application.

**Comment by remi-braun** [comment]
My use case is that the user can give a path and this path can be a URL.

I just need this URL recognized as one (and not as a classic path) in order not to fail when asking `is_dir` or `is_file`.
(it would be even better to implement all these functions but I think this is beyond the scope your lib, maydbe [this lib](https://pypi.org/project/urlpath/) can be useful ? )

For now I'm using the [`validators`](https://validators.readthedocs.io/en/latest/) lib, but this adds a dependency and another test

**Comment by pjbull** [comment]
Gotcha. I think the only URLs that we would scope to handle are links to files that are actually on one of the cloud providers, not generic URLs.

For generic URLs, we usually check with [`urllib.parse`](https://docs.python.org/3/library/urllib.parse.html) first since it's robust and standard library.

**Comment by pjbull** [comment]
This just got mentioned in #252.

One of the big things missing there is the research to track down documentation of the potential URL formats. Even just for S3, there are a fair number of possible URLs that need to parse the bucket information and potentially the credentials (e.g., if they are presigned) from the URL. We'll need a better understanding of what the burden here looks like before we decide to implement and support.

**Comment by remi-braun** [comment]
My usecase is (not that) simple: I am using `cloudpathlib` to unify the path management in my library.

The problem is that a S3 file can be described by a S3Path or by an URL.
I would be very grateful if those URLs could be automatically routed to a S3Path rather than testing on my side if the provided string is an URL (and convert it myself to a S3Path) 

However, I am not aware of the complexity of the different S3 URLs so this may be a bit naive 😅

**Comment by pjbull** [comment]
> URLs could be automatically routed to a S3Path

Yep, what you are describing is the feature that is tracked in this issue. 

> complexity of the different S3 URLs

From what we've seen, these URLs formats have changed over time and at least in some points could be different from region-to-region.

If someone can do the legwork to track down the canonical references for URLs for the different providers, we can look at pulling together the implementation.

**Comment by remi-braun** [comment]
I found [this](http://www.wryway.com/blog/aws-s3-url-styles/) blog post if it can help 😄  

Maybe this [lib](https://github.com/marazmiki/s3-parse-url) can help ?

Personnaly, my URLs (from S3-compatible Ceph storage) look like: `https://s3.endpoint/bucket/file.extension`
